Allow disabling of Iceberg table statistics

piyushn-stripe opened the pull request:
This change adds support to wrap the lookup of Iceberg's table statistics behind a config flag / session property. This gets the Iceberg connector to a similar state as the Hive connector regarding stats retrieval (as we gate that behind `hive.table-statistics-enabled`). 
cc @findepi / @raunaqmorarka - as we discussed this a few days back on the Trino slack channel.

piyushn-stripe commented:
p.s. I think folks on our end (Stripe) should have sent over the signed CLA to `cla@trino.io`

raunaqmorarka commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergConfig.java:
```diff
@@ -91,4 +92,22 @@ public IcebergConfig setMaxPartitionsPerWriter(int maxPartitionsPerWriter)
     }
+
+    /**
+     * In case of some queries / tables, retrieving table statistics from Iceberg
```
This probably belongs in docs

findepi replied:
It's fine to have this as a code comment as well, but that's not a javadoc. A javadoc should describe what a thing is, but that's not needed here, as the config is self-explanatory.

Please convert it into `/*` or `//` ordinary comment.

piyushn-stripe replied:
Sure, can update to a usual comment. I'll also add a note on this property in the iceberg docs.

findepi replied:
Let's make docs a followup, as there will surely be some back and forth on the wording.
Actually, i wouldn't document that, as this is an escape hatch only, and we should rather fix the underlying problem(s) instead.
For example, `hive.table-statistics-enabled` remains undocumented.

raunaqmorarka commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergConfig.java:
```diff
@@ -91,4 +92,22 @@ public IcebergConfig setMaxPartitionsPerWriter(int maxPartitionsPerWriter)
+     * can take 20+ seconds. This config allows the user / operator the option
+     * to opt out of retrieving table statistics in those cases to speed up query planning.
+     */
+    @Config("iceberg.table-statistics-enabled")
```
Let's make it consistent with the session property skip the `table`

findepi replied:
This is named consistently with hive's `hive.table-statistics-enabled`

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSessionProperties.java:
```diff
@@ -61,6 +61,7 @@
     private static final String PARQUET_MAX_READ_BLOCK_SIZE = "parquet_max_read_block_size";
     private static final String PARQUET_WRITER_BLOCK_SIZE = "parquet_writer_block_size";
     private static final String PARQUET_WRITER_PAGE_SIZE = "parquet_writer_page_size";
+    private static final String STATISTICS_ENABLED = "statistics_enabled";
```
This is consistent with hive's `statistics_enabled`, but indeed not consistent with property name.

i would go with current naming as proposed in this PR.

findepi reviewed (commented):
Please change the commit and PR title to something like

```
Allow to disable Iceberg table statistics
```

findepi commented:
> p.s. I think folks on our end (Stripe) should have sent over the signed CLA to `cla@trino.io`

@piyushn-stripe for a company-sponsored contribution, there company needs to sign the corporate CLA
and the person contributing needs to sign the individual CLA. Please make sure both are sent.

cc @martint

piyushn-stripe commented:
@findepi / @raunaqmorarka - updated based on feedback. 
I also just sent over the CLAs to cla@trino.io.

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSessionProperties.java:
```diff
@@ -182,6 +183,11 @@ public IcebergSessionProperties(
+                    STATISTICS_ENABLED,
+                    "Expose table statistics",
+                    icebergConfig.isTableStatisticsEnabled(),
+                    false))
```
i noticed the indentation here is not the same as on then lines above.
you can get formatter definition from https://github.com/airlift/codestyle/

piyushn-stripe replied:
yes, I was missing the airflift style file and using something we have internally. Fixed

piyushn-stripe commented:
@findepi - do the changes look good to you? I can rebase and get this ready for a merge if so.

findepi commented:
@piyushn-stripe the code LGTM. The CLA bot is not satisfied yet though.

piyushn-stripe commented:
@findepi - can we re-trigger it? I got a note from Martin yesterday indicating I've been added to the Trino community post signing the CLA.

phd3 commented:
@cla-bot check

phd3 commented:
@piyushn-stripe doesn't look like it has gone through yet on the trinodb/CLA repo. cc @martint

martint commented:
Let me check that. I may have forgotten to push the commits with all the CLA additions. The CLA is processed, though, and it’s good to go.

piyushn-stripe commented:
@cla-bot check

piyushn-stripe commented:
@findepi - if you have cycles, could you take a look? Dropped the ball on this with the failing CLA checks (now fixed). I have rebased so this should be ready to go.

alexjo2144 commented:
@raunaqmorarka do you think this should also disable using statistics for DF? https://github.com/trinodb/trino/blob/master/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java#L129

raunaqmorarka commented:
> @raunaqmorarka do you think this should also disable using statistics for DF? https://github.com/trinodb/trino/blob/master/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java#L129

That depends on whether the inclusion of column stats is still a huge penalty. Back when this was originally discussed the TableStatisticsMaker code was throwing some errors and those are apparently fixed now https://trinodb.slack.com/archives/CGB0QHWSW/p1627048943120500?thread_ts=1626836724.048000&cid=CGB0QHWSW

I think we can skip including column stats when dynamic filter is EMPTY
Also, is there any way we can request the inclusion of column stats for only a subset of columns through the iceberg API ? I think in our case we're only interested in column stats for DynamicFIlter#getColumnsCovered

findepi commented:
> That depends on whether the inclusion of column stats is still a huge penalty. Back when this was originally discussed the TableStatisticsMaker code was throwing some errors and those are apparently fixed now https://trinodb.slack.com/archives/CGB0QHWSW/p1627048943120500?thread_ts=1626836724.048000&cid=CGB0QHWSW

https://github.com/trinodb/trino/pull/8645 was not enough. 
there was also https://github.com/trinodb/trino/pull/9719, but still https://github.com/trinodb/trino/issues/9716 remains.
@alexjo2144  is working on that


> I think we can skip including column stats when dynamic filter is EMPTY Also, is there any way we can request the inclusion of column stats for only a subset of columns through the iceberg API ? I think in our case we're only interested in column stats for DynamicFIlter#getColumnsCovered

good idea. Not something tbd in this PR though.

> do you think this should also disable using statistics for DF?

@alexjo2144 very good question, i didn't think about it.

these are technically "file statistics" and "table statistics" is something derived from "file statistics"

- if we wanted to disable both, toggle would need a different name (ignore file statistics)
- what would be the rationale? we disable "table statistics" because they can be slow to compute, so we improve query perf in some cases, and prevent some failures (like https://github.com/trinodb/trino/pull/9719). Do we have evidence that disabling file stats during split generation improves performance, or avoids failures?

i think we should not change IcebergSplitSource
